Some SV-COMP fixes

fbrausse opened the pull request:
Changes made by this PR:
1. introduce type-compat assertion in constructor of `if2t`, cf. #204 and rejects code-to-data and data-to-code deferences
   - fixes crashes related to dereferencing `code` as `data` and vice-versa, e.g. in `ldv-linux-3.0/module_get_put-drivers-net-wan-farsync.ko.cil.out.i` (still wrong verdict, though), see fbrausse#2 for more instances related to this
   - makes `regression/llvm/ternary` CORE->KNOWNBUG: this assertion triggers for https://github.com/esbmc/esbmc/blob/20c2beaf6d2ee2f3f2fe3686ae073de9e4e41cec/regression/llvm/ternary/main.c#L19 in https://github.com/esbmc/esbmc/blob/20c2beaf6d2ee2f3f2fe3686ae073de9e4e41cec/src/util/migrate.cpp#L909 *Edit:* Added a fix to this PR.
2. do not exclude call targets based on their signature at all anymore in goto-symex, *edit*: instead assert when they are known to be wrong; basically following up on #582
   - makes `regression/esbmc-unix2/11_lu-fig4-complete.fixed` CORE->KNOWNBUG (even though this should've been a known bug before): pthread_create on https://github.com/esbmc/esbmc/blob/20c2beaf6d2ee2f3f2fe3686ae073de9e4e41cec/regression/esbmc-unix2/11_lu-fig4-complete.fixed/lu-fig4-complete.fixed.c#L33-L34 was considered an "illegal target" before (e.g., try to insert an `assert(0)` at the top of this function - it's not found in current master), now it is a valid target - until the function-call actually happens. I suppose we should produce something like `derefence-failure` at that point. This PR ~does not yet do that~ does do that now.
3. fix (some cases of?) the rvalue-to-array errors by stitching together the new array from the old in case it is a byte-array (do we maybe want a new type for that - e.g. "in-memory-representation"? It seems related to how unions are handled now). Related issues: #300, #134.

I am running the SV-COMP benchmarks over this now.

fbrausse committed:
[WIP] fix crash in sv-comp ldv-linux-3.0/module_get_put-drivers-net-wan-farsync.ko.cil.out.i -- still wrong verdict

fbrausse committed:
[dereference] restructure check_code_access()

fbrausse committed:
[dereference] restructure check_code_access()

fbrausse committed:
[symex] function: do not exclude call targets based on their signature

fbrausse committed:
[symex] function: do not exclude call targets based on their signature

fbrausse committed:
[dereference] fix rvalue-to-array construction

fbrausse committed:
[dereference] fix rvalue-to-array construction

fbrausse commented:
Regarding 3: I was bringing up the question "maybe a new type?" because in `dereferencet` all knowledge over whether something was a union to begin with seems to be lost. That's why the patch is not touching the general `build_reference_to_rec()` but only the special case `construct_struct_ref_from_const_offset_array()`. Maybe instead of a new type2t it is enough to extend `is_union_type()`? Though probably things start breaking again, then...

fbrausse committed:
[regression] mark esbmc-unix2/11_lu-fig4-complete.fixed KNOWNBUG

This test case has the problem that a function with 4 parameters is expected
to be run where only 1 parameter is available - and even its type differs
incompatibly (int vs. void *).

fbrausse commented:
Master 20c2beaf6d2ee2f3f2fe3686ae073de9e4e41cec 6GB, 5min:
```
Statistics:          15457 Files
  correct:            8295
    correct true:     5156
    correct false:    3139
  incorrect:            72
    incorrect true:     40
    incorrect false:    32
  unknown:            7090
  Score:             11659 (max: 25029)
```
This PR efb34b7e0245f7a5a8172d76907ece6f23879ce0 6GB, 5min:
```
Statistics:          15457 Files
  correct:            8312
    correct true:     5164
    correct false:    3148
  incorrect:            75
    incorrect true:     42
    incorrect false:    33
  unknown:            7070
  Score:             11604 (max: 25029)
```

fbrausse commented:
Diffs ~(w/o logs)~:
- [no-overflow](https://fbrausse.github.io/esbmc/svcomp/gh-606/no-overflow/results.2022-01-24_10-14-39.diff.html#/)
- [termination](https://fbrausse.github.io/esbmc/svcomp/gh-606/termination/results.2022-01-24_10-14-40.diff.html#/)
- [unreach-call](https://fbrausse.github.io/esbmc/svcomp/gh-606/unreach-call/results.2022-01-24_10-14-41.diff.html#/)
- [valid-memsafety](https://fbrausse.github.io/esbmc/svcomp/gh-606/valid-memsafety/results.2022-01-24_10-14-50.diff.html#/)

lucasccordeiro commented:
@fbrausse: these results are very encouraging! 

Can I ask you whether you could take a look at the cases we now provide incorrect results?

Overall, this PR looks good since it solves more verification tasks than our master branch.

fbrausse commented:
new incorrect results:
- unreach-call:
  - (1) TIMEOUT -> false(unreach-call): [log](https://fbrausse.github.io/esbmc/svcomp/gh-606/new/results/esbmc-kind.2022-01-21_12-06-07.logfiles/SV-COMP22_unreach-call.32_1_cilled_ok_nondet_linux-3.4-32_1-drivers--media--video--cx231xx--cx231xx-dvb.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.yml.log); was already [TIMEOUT](https://sv-comp.sosy-lab.org/2022/results/results-verified/esbmc-kind.2021-12-07_20-41-05.logfiles/SV-COMP22_unreach-call.32_1_cilled_ok_nondet_linux-3.4-32_1-drivers--media--video--cx231xx--cx231xx-dvb.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.yml.log) in SV-COMP22
  - unknown -> true:
    - (2) c/ldv-linux-3.0/module_get_put-drivers-net-wan-farsync.ko.cil.out.yml: [was](https://fbrausse.github.io/esbmc/svcomp/gh-606/old/results/esbmc-kind.2022-01-22_15-08-31.logfiles/SV-COMP22_unreach-call.module_get_put-drivers-net-wan-farsync.ko.cil.out.yml.log) segfault in `tuple_node_smt_ast::make_free(smt_convt *)` in master and in [SV-COMP22](https://sv-comp.sosy-lab.org/2022/results/results-verified/esbmc-kind.2021-12-07_20-41-05.logfiles/SV-COMP22_unreach-call.module_get_put-drivers-net-wan-farsync.ko.cil.out.yml.log)
    - (3) c/ldv-linux-3.16-rc1/43_2a_bitvector_linux-3.16-rc1.tar.xz-43_2a-drivers--media--platform--timblogiw.ko-entry_point.cil.out.yml: [was](https://fbrausse.github.io/esbmc/svcomp/gh-606/old/results/esbmc-kind.2022-01-22_15-08-31.logfiles/SV-COMP22_unreach-call.43_2a_bitvector_linux-3.16-rc1.tar.xz-43_2a-drivers--media--platform--timblogiw.ko-entry_point.cil.out.yml.log) `Can't construct rvalue reference to array type during dereference` in master and in [SV-COMP22](https://sv-comp.sosy-lab.org/2022/results/results-verified/esbmc-kind.2021-12-07_20-41-05.logfiles/SV-COMP22_unreach-call.43_2a_bitvector_linux-3.16-rc1.tar.xz-43_2a-drivers--media--platform--timblogiw.ko-entry_point.cil.out.yml.log)
    - (4) c/ldv-validator-v0.8/linux-stable-1dfa93a-1-100_1a-drivers--usb--serial--kobil_sct.ko-entry_point_ldv-val-v0.8.cil.out.yml: [was](https://fbrausse.github.io/esbmc/svcomp/gh-606/old/results/esbmc-kind.2022-01-22_15-08-31.logfiles/SV-COMP22_unreach-call.linux-stable-1dfa93a-1-100_1a-drivers--usb--serial--kobil_sct.ko-entry_point_ldv-val-v0.8.cil.out.yml.log) `Can't construct rvalue reference to array type during dereference` in master and in [SV-COMP22](https://sv-comp.sosy-lab.org/2022/results/results-verified/esbmc-kind.2021-12-07_20-41-05.logfiles/SV-COMP22_unreach-call.linux-stable-1dfa93a-1-100_1a-drivers--usb--serial--kobil_sct.ko-entry_point_ldv-val-v0.8.cil.out.yml.log)
- valid-memsafety:
  - TIMEOUT -> false(valid-memtrack):
    - (5) c/Juliet_Test/CWE401_Memory_Leak---s03---CWE401_Memory_Leak__struct_twoIntsStruct_calloc_45_bad.yml: [is now](https://fbrausse.github.io/esbmc/svcomp/gh-606/new/results/esbmc-kind.2022-01-21_12-06-07.logfiles/SV-COMP22_valid-memsafety.CWE401_Memory_Leak---s03---CWE401_Memory_Leak__struct_twoIntsStruct_calloc_45_bad.yml.log) `dereference failure: forgotten memory: dynamic_1602_array`; was [unknown](https://sv-comp.sosy-lab.org/2022/results/results-verified/esbmc-kind.2021-12-07_20-41-05.logfiles/SV-COMP22_valid-memsafety.CWE401_Memory_Leak---s03---CWE401_Memory_Leak__struct_twoIntsStruct_calloc_45_bad.yml.log) in SV-COMP22. I've used the `esbmc-wrapper.py` from `master` for these runs, it differs from the SV-COMP22 one. Since the verdict is the same now and in SV-COMP22, this may be a bug in the wrapper script.

old incorrect results:
- (6) true -> unknown: c/ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--media--usb--ttusb-budget--dvb-ttusb-budget.ko.cil.yml [aborting](https://fbrausse.github.io/esbmc/svcomp/gh-606/new/results/esbmc-kind.2022-01-21_12-06-07.logfiles/SV-COMP22_unreach-call.linux-4.0-rc1---drivers--media--usb--ttusb-budget--dvb-ttusb-budget.ko.cil.yml.log) here with this PR: https://github.com/esbmc/esbmc/blob/96a38dcfa89b2df70ce2daa99607c36d8acbb695/src/goto-symex/symex_function.cpp#L285-L296 will still crash even when filtering out call targets based on the same condition in the `if` above:
  ```
  src/goto-symex/goto_symex_state.h:329: void goto_symex_statet::pop_frame(): Assertion `call_stack.back().goto_state_map.size() == 0' failed.
  ```
  *edit:* this was already [incorrectly true](https://sv-comp.sosy-lab.org/2022/results/results-verified/esbmc-kind.2021-12-07_20-41-05.logfiles/SV-COMP22_unreach-call.linux-4.0-rc1---drivers--media--usb--ttusb-budget--dvb-ttusb-budget.ko.cil.yml.log) in SV-COMP22 as well
- (7) false(valid-memtrack) -> TIMEOUT: Juliet_Test/CWE401_Memory_Leak---s01---CWE401_Memory_Leak__int64_t_calloc_45_bad.yml; was incorrectly [false(valid-memtrack)](https://fbrausse.github.io/esbmc/svcomp/gh-606/old/results/esbmc-kind.2022-01-22_15-08-31.logfiles/SV-COMP22_valid-memsafety.CWE401_Memory_Leak---s01---CWE401_Memory_Leak__int64_t_calloc_45_bad.yml.log) in master; [unknown](https://sv-comp.sosy-lab.org/2022/results/results-verified/esbmc-kind.2021-12-07_20-41-05.logfiles/SV-COMP22_valid-memsafety.CWE401_Memory_Leak---s01---CWE401_Memory_Leak__int64_t_calloc_45_bad.yml.log) in SV-COMP22; and is TIMEOUT with this PR. To me this suggests a problem with `esbmc-wrapper.py` as in (5) above: verdict for master and SV-COMP22 the same, but interpretation different.

fbrausse commented:
My take on these results:
- (5) and (7) above seem to be caused by a different interpretation in `esbmc-wrapper.py`.
- (6) crashing due to these changes is probably better than a wrong result, will investigate further whether for instance adding a guard condition to the function call makes it produce the correct result
- (2), (3) and (4) are hard to say - they have been crashing before and give wrong results now, will take a look at whether the `rvalue-to-array` fix has anything to do with it
- (1) definitely requires looking into more

lucasccordeiro commented:
@rafaelsamenezes and @mikhailramalho: do you have any thoughts?

mikhailramalho commented on src/goto-symex/symex_function.cpp:
```diff
@@ -283,8 +283,9 @@ void goto_symext::symex_function_call_code(const expr2tc &expr)
-    to_code_type(tmp_type).arguments.size() != arguments.size() &&
-    !to_code_type(tmp_type).ellipsis)
+    to_code_type(tmp_type).arguments.size() > arguments.size() ||
+    (0 && to_code_type(tmp_type).arguments.size() < arguments.size() &&
```
`0` forces the second part of the guard to be false, can you remove it?

fbrausse replied:
Thanks, done. I've also reworded the error message.

mikhailramalho commented on src/goto-symex/symex_function.cpp:
```diff
@@ -412,7 +413,7 @@ void goto_symext::symex_function_call_deref(const expr2tc &expr)
     // C++ does not allow declarations without prototypes
     // TODO: see GitHub issues #584 and #585
-    if(ct.arguments.size() && ct.arguments.size() != call.operands.size())
+    if(0 && ct.arguments.size() && ct.arguments.size() != call.operands.size())
```
`0` forces the guard to be false, can you remove it?

Shouldn't it be `ct.arguments.size() > call.operands.size()` like the previous one?

fbrausse replied:
I'm feeling uneasy about this `illegal_filter`: why is anything illegal at this point in symex? If not a valid call target, shouldn't it be guarded by some predicate? For instance, if all targets were "illegal" that would mean execution cannot continue - this should be VERIFICATION FAILED, no?

fbrausse replied:
~For now we are again filtering, but I'm checking whether this filter could actually be a no-op and if not why not. Note that this constitutes a semantic change in the patch and the SV-COMP numbers may not be accurate anymore.~

*Edit*: I have found a better solution.

mikhailramalho reviewed (approved):
LGTM!

I think it's fine to merge PRs that introduce FP and FN, as long as they were crashes or timeouts previously. I also suggest that the fixes should come in another PR.

lucasccordeiro reviewed (approved):
Apart from the issues pointed out by @mikhailramalho, this PR looks good ;-)

Many thanks for submitting this patch, @fbrausse!

lucasccordeiro commented:
@fbrausse: Once you address @mikhailramalho's comments, could you please rebase this branch?

fbrausse commented:
> Once you address @mikhailramalho's comments, could you please rebase this branch?

@lucasccordeiro done.

fbrausse committed:
[regression] add copy of esbmc-unix2/11_lu-fig4-complete.fixed showing that call target is not considered

fbrausse commented:
Re-inserting the filter based on #params vs. #args makes the (undefined) `regression/esbmc-unix2/11_lu-fig4-complete.fixed` VERIFICATION SUCCESSFUL again. I've added a KNOWNBUG copy showing that the call target is not considered ("illegal") by ESBMC.

fbrausse commented:
Note that re-inserting the filter constitutes a semantic change in the patch and the SV-COMP numbers may not be accurate anymore. I'd rather assign semantics to the case when the parameters in the declaration to not match the call.

fbrausse committed:
[goto-symex] function call: do not filter invalid targets, instead assert it when known

fbrausse commented:
Sorry for the confusion: I have found a better solution: instead of silently dropping targets that have been generated, we add a `claim()` in case not enough arguments have been provided to a call during symex. If my understanding of the `claim()` procedure is correct, this should not change the semantics, but instead give a useful `Violated property: [...] function call: not enough arguments` and VERIFICATION FAILED. Do you think it makes sense to re-run this over SV-COMP?

lucasccordeiro commented:
> Note that re-inserting the filter constitutes a semantic change in the patch and the SV-COMP numbers may not be accurate anymore.

@fbrausse: shall we rerun this patch over the SV-COMP benchmarks before we merge it?

lucasccordeiro commented:
>Do you think it makes sense to re-run this over SV-COMP?

I would say so.

@mikhailramalho and @rafaelsamenezes: Do you have any thoughts here?

fbrausse commented:
> I would say so.

Will do. I'll go back to before rebasing so we can compare the numbers.

mikhailramalho commented:
I also find it strange to have that check there, I thought the frontend
would complain about it way before symex. I think it's fine to run it on
SV-Comp, but I don't think this check should be enabled by default.

Maybe this was a sanity check for when we implemented c++ default
arguments? IIRC, we copy the default arguments to the callee irep.


Em seg., 24 de jan. de 2022 às 13:01, Lucas Cordeiro <
***@***.***> escreveu:

> Do you think it makes sense to re-run this over SV-COMP?
>
> I would say so.
>
> @mikhailramalho <https://github.com/mikhailramalho> and @rafaelsamenezes
> <https://github.com/rafaelsamenezes>: Do you have any thoughts here?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/606#issuecomment-1020326236>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH3HFCYFXGWVZFVQHJDUXWAXXANCNFSM5MPIZ7OQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.

rafaelsamenezes reviewed (approved):
I agree with @mikhailramalho. If the new FP/FN are due to ESBMC now being able to process more files, then we should accept it and focus on fixing/identifying the new issues.

fbrausse commented:
>  but I don't think this check should be enabled by default.

I agree. If SV-COMP runs through nicely, I'll make it an `std::all_of()` assertion. *Edit* You are talking about the (ex-)`illegal_filter`, right? Because the new `claim()` does indeed trigger, e.g. for `regression/esbmc-unix2/11_lu-fig4-complete*/`.

fbrausse commented:
The Windows CI is failing because it is built w/o `assert()`, see item 1. in the initial post of this PR.

rafaelsamenezes commented:
> The Windows CI is failing because it is built w/o assert(), see item 1. in the initial post of this PR.

I am not sure how to handle this. If the assert is a requirement, then we should convert to a throw/abort. If is just a sanity check, then we probably should mark the test as a THROUGHOUT.

fbrausse commented:
We have about 1 day to figure out where this conversion from bool to float is being performed after migrate ;)

fbrausse committed:
[goto-programs] sideeffects: typecast true-op in "cond ? : other" to expression's type

fbrausse commented:
This PR (56be87fcf1118ce38711e6b3200256225bc86b4c):
```
Statistics:          15457 Files
  correct:            8314
    correct true:     5164
    correct false:    3150
  incorrect:            76
    incorrect true:     42
    incorrect false:    34
  unknown:            7067
  Score:             11590 (max: 25029)
```

lucasccordeiro commented:
Thanks, @fbrausse.

Could you please rebase this branch?

fbrausse commented:
> Thanks, @fbrausse.
> 
> Could you please rebase this branch?

@lucasccordeiro done
